keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,886 @@
+    // there are no conflicts with other CSI drivers on the cluster, the recommendation
+    // is to use sp-<uuid>.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	Spec   CSIStoragePoolSpec
+	Status CSIStoragePoolStatus
+}
+
+// CSIStoragePoolSpec contains the constant attributes of a CSIStoragePool.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// NodeList, but not both. If neither is set, the pool is assumed
+// to be available in the entire cluster.
+type CSIStoragePoolSpec struct {
+    // The CSI driver that provides access to the storage pool.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+	// NodeTopology can be used to describe a storage pool that is available
+	// only for nodes matching certain criteria.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// Nodes can be used to describe a storage pool that is available
+	// only for certain nodes in the cluster.
+	//
+	// +listType=set
+	// +optional
+	Nodes []string
+}
+
+// CSIStoragePoolStatus contains runtime information about a CSIStoragePool.
+//
+// It is expected to be extended with other
+// attributes which do not depend on the storage class, like health of
+// the pool. Therefore it has the list of
+// `CSIStorageByClass` instances instead of just the capacity
+// and the storage class being in the spec.
+type CSIStoragePoolStatus struct {
+	// Some information, like the actual usable capacity, may
+	// depend on the storage class used for volumes.
+	//
+	// +patchMergeKey=storageClassName
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=storageClassName
+	// +optional
+	Classes []CSIStorageByClass `patchStrategy:"merge" patchMergeKey:"storageClassName" json:"classes,omitempty" protobuf:"bytes,4,opt,name=classes"`
+}
+
+// CSIStorageByClass contains information that applies to one storage
+// pool of a CSI driver when using a certain storage class.
+type CSIStorageByClass struct {
+	// The storage class name matches the name of some actual
+	// `StorageClass`, in which case the information applies when
+	// using that storage class for a volume. There are also two
+	// special names:
+	// - <ephemeral> for storage used by ephemeral inline volumes (which
+	//   don't use a storage class)
+	// - <fallback> for storage that is the same regardless of the storage class;
+	//   it is applicable if there is no other, more specific entry
+	StorageClassName string `json:"storageClassName" protobuf:"bytes,1,name=storageClassName"`
+
+	// Capacity is the size of the largest volume that currently can
+	// be created. This is a best-effort guess and even volumes
+	// of that size might not get created successfully.
+	// +optional
+	Capacity *resource.Quantity `json:"capacity,omitempty" protobuf:"bytes,2,opt,name=capacity"`
```

Comment:
Here, only storing `the largest volume that currently can be created` field will cause conflicts easily, since storage drivers need some time to update it, but scheduler works more faster and will get the wrong value if this field is not updated in time.
And i agree with you, only storing `capacity` filed here is problematic too.
Can we keep both `capacity` and `largest volume that currently can be created` here, and take both of them into account when scheduling ? If so, we can reduce the risk of race conditions in some scenarios.

Reply:
> storage drivers need some time to update it

That depends a bit. For PVCs, external-provisioner can update the field after each `CreateVolume`. It's harder for ephemeral volumes because then it has to poll.

> but scheduler works more faster and will get the wrong value if this field is not updated in time.

True, the scheduler may end up scheduling several pods at once and then will have to make decisions based on capacity information that has not been updated yet.

> Can we keep both capacity and largest volume that currently can be created here, and take both of them into account when scheduling ? If so, we can reduce the risk of race conditions in some scenarios.

What are those scenarios, and how would that work?

We can add more fields here and make them informational, like "raw capacity". But what's the semantic of that then? Where's that value supposed to come from? I suspect it's going to be something that's vendor-specific and thus not something that the default Kubernetes scheduler will be able to use. It's a typical apple and orange comparison: just because you have 1 kg of apples that doesn't mean that you can convert them into 1 kg of oranges.

BTW, we don't even need this to be part of the documented API. Because `CSIStoragePool` is now a stand-alone object, it supports annotations. A vendor could store its "raw capacity" there.

Reply:
> It's a typical apple and orange comparison: just because you have 1 kg of apples that doesn't mean that you can convert them into 1 kg of oranges.

A better analogy is comparing apples against apple juice: 1 kg of apples does not yield 1 kg of apple juice... it depends on the apples, the manufacturing process, etc.

Reply:
This is used as a hint right now. We require the scheduler must tolerate provisioning failures but don't ask the storage driver to report the correct consumable capacity.

There are some considerations:

1) even if it's the correct consumable capacity, but we don't have total capacity now, the information may be incorrect if there are some in-flight volume operations. We need total capacity information then it's possible to make deterministic scheduling decisions (like pod scheduling based on node total allocated CPU), this is [under discussion](https://github.com/container-storage-interface/spec/issues/301) 
2) unlike CPU/Memory, raw capacity normally does not equal to the usage capacity (what the user requested). It depends on storage implementation. The scheduler (volume scheduling) can't precompute the raw capacity based on the requested capacity for different drivers. The driver can report recomputed usage capacity but storage implementations may have a fixed ratio between `<usage>:<consumed>`, some may not (filesystems are more complicated).
3) the consumable capacity may change when disks are added/removed in the storage backend.

So no matter what, the consumable capacity left may change. 

The storage driver can try it's best to report recomputed usable capacity when the storage implementation has a fixed ratio between `<usage>:<consumed>` capacities. This can avoid unnecessary failures when the capacity is near full. In the future, when the CSI API can report total capacity, this type of storage implementation can report total capacity which the volume scheduling can take into account and make deterministic scheduling decisions like CPU/Memory.

[This doc](https://docs.google.com/document/d/1WtX2lRJjZ03RBdzQIZY3IOvmoYiF5JxDX35-SsCIAfg) has some discussions about this.

Reply:
It's not clear to me how a driver determine what is the "largestVolumeCapacity". The "Available" and "Total" capacity fields in [this doc](https://docs.google.com/document/d/1WtX2lRJjZ03RBdzQIZY3IOvmoYiF5JxDX35-SsCIAfg/edit#) make more sense and the scheduler can make decisions based on that.

Reply:
The scheduler cannot assume that a volume of size `X` can be created just because `X <= Available`. That completely ignores additional constraints like fragmentation. Basically the definition of "available" is too vague for it to be useful and therefore "maximum volume size" is used instead.

OTOH, if that comparison is good enough for a driver, then its fine to just use "largestVolumeCapacity = available capacity".

But the idea is that a driver can do better than that by also taking into account other constraints. For example, a LVM driver might return the maximum of the free space in all the volume groups instead of the sum of those values. 

The CSI spec for `GetCapacityResponse` just says that it contains "available capacity, in bytes, of the storage that can be used to provision volumes" and indicates that this "SHALL take those [capabilities and parameters] into
consideration when calculating the available capacity of the storage".

At first glance that sounds more like "return the sum of all possible volumes size", but I think it's okay for a driver to return the maximum volume size instead.

Reply:
> The "Available" and "Total" capacity fields in this doc make more sense

Regarding "total capacity": this had been proposed in https://github.com/container-storage-interface/spec/issues/301 and the CSI community pretty much rejected the idea as unsuitable for the intended purpose. I tend to agree.

Reply:
Done, see 0fd7bc844306629eff5b9f2ea25ca25d212ff15d.

Reply:
Regarding the freshness of storage capacity status, it is not only how fast the updated capacity propagates to the scheduler, but also how often the provisioner is going to pull this information via ```GetCapacity```. 

Note that a scheduling cycle can be as low as 10ms, and we continue to bring it down, so I feel that without proper caching of the capacity in the scheduler (like we do for node resource like cpu and memory), the scheduler may end up making decisions on invalid capacity information more often than not, and so limiting the impact of this whole proposal.